File: ch04/WKWebView/WKWebView/ViewController.m
//
//  ViewController.m
//  WKWebView
//
//  Created by MyZiyue on 2018/12/12.
//  Copyright © 2018 MyZiyue.com. All rights reserved.
//

#import "ViewController.h"
#import <WebKit/WebKit.h>

@interface ViewController () <WKNavigationDelegate>

@property (nonatomic, strong) WKWebView* webView;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    CGRect screen = [[UIScreen mainScreen] bounds];
    
    // 按键栏
    // 按键栏高
    CGFloat buttonBarWidth = 316;
    UIView* buttonBar = [[UIView alloc] initWithFrame:CGRectMake((screen.size.width - buttonBarWidth)/2, 20, buttonBarWidth, 30)];
    
    [self.view addSubview:buttonBar];
    
    // 1. 添加LoadHTMLString按钮
    UIButton* buttonLoadHTMLString = [UIButton buttonWithType:UIButtonTypeSystem];
    [buttonLoadHTMLString setTitle:@"LoadingHTMLSTring" forState:UIControlStateNormal];
    buttonLoadHTMLString.frame = CGRectMake(0, 0, 117, 30);
    // 指定事件处理方法
    [buttonLoadHTMLString addTarget:self action: @selector(testLoadHTMLString:) forControlEvents:UIControlEventTouchUpInside];
    [buttonBar addSubview:buttonLoadHTMLString];
    
    // 2. 添加LoadDatan按钮
    UIButton* buttonLoadData = [UIButton buttonWithType:UIButtonTypeSystem];
    [buttonLoadData setTitle:@"LoadData" forState:UIControlStateNormal];
    buttonLoadData.frame = CGRectMake(137, 0, 67, 30);
    // 指定事件处理方法
    [buttonLoadData addTarget:self action:@selector(testLoadData:) forControlEvents:UIControlEventTouchUpInside];
    [buttonBar addSubview:buttonLoadData];
    
    // 3. 添加LoadRequest按钮
    UIButton* buttonLoadRequest = [UIButton buttonWithType:UIButtonTypeSystem];
    [buttonLoadRequest setTitle:@"LoadRequest" forState:UIControlStateNormal];
    buttonLoadRequest.frame = CGRectMake(224, 0, 92, 30);
    // 指定事件处理方法
    [buttonLoadRequest addTarget:self action:@selector(testLoadRequest:) forControlEvents:UIControlEventTouchUpInside];
    [buttonBar addSubview:buttonLoadRequest];
    
    // 4. 添加WKWebView
    self.webView = [[WKWebView alloc] initWithFrame:CGRectMake(0, 60, screen.size.width, screen.size.height-80)];
    [self.view addSubview:self.webView];
}

-(void)testLoadHTMLString:(id)sender{
    NSString *htmlPath = [[NSBundle mainBundle] pathForResource:@"index" ofType:@"html"];
    NSURL *bundleUrl = [NSURL fileURLWithPath:[[NSBundle mainBundle] bundlePath]];
    NSError *error = nil;
    
    NSString *html = [[NSString alloc] initWithContentsOfFile:htmlPath encoding:NSUTF8StringEncoding error:&error];
    
    if (error == nil){
        [self.webView loadHTMLString:html baseURL:bundleUrl];
    }
}

-(void)testLoadData:(id)sender {
    NSString *htmlPath = [[NSBundle mainBundle] pathForResource:@"index" ofType:@"html"];
    NSURL *bundleUrl = [NSURL fileURLWithPath:[[NSBundle mainBundle] bundlePath]];
    NSData *htmlData = [[NSData alloc] initWithContentsOfFile:htmlPath];
    
    [self.webView loadData:htmlData MIMEType:@"text/html" characterEncodingName:@"UTF-8" baseURL:bundleUrl];
}

-(void)testLoadRequest:(id)sender {
    NSURL *url = [NSURL URLWithString:@"https://www.baidu.com"];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    [self.webView loadRequest:request];
    self.webView.navigationDelegate = self;
}

// 实现WKNavigationDelegate委托协议
// 开始加载时调用
-(void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation{
    NSLog(@"开始加载");
}
// 当内容开始返回时调用
-(void)webView:(WKWebView *)webView didCommitNavigation:(WKNavigation *)navigation {
    NSLog(@"内容开始返回");
}
// 加载完成之后调用
-(void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    NSLog(@"加载完成");
}
// 加载失败时调用
-(void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    NSLog(@"加载失败 error : %@", error.description);
}
@end

File: ch04/SWKWebView/SWKWebView/ViewController.swift
//
//  ViewController.swift
//  SWKWebView
//
//  Created by MyZiyue on 2018/12/12.
//  Copyright © 2018 MyZiyue.com. All rights reserved.
//

import UIKit
import WebKit

class ViewController: UIViewController,WKNavigationDelegate {
    var webView: WKWebView!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let screen = UIScreen.main.bounds
        
        // 按钮栏
        // 按钮栏宽
        let buttonBarWidth:CGFloat = 316
        let buttonBar = UIView(frame: CGRect(x:(screen.size.width - buttonBarWidth)/2, y: 20, width:buttonBarWidth, height:30))
        self.view.addSubview(buttonBar)
        
        // 1. 添加LoadHTML String按钮
        let buttonLoadHTMLString = UIButton(type: UIButton.ButtonType.system)
        buttonLoadHTMLString.setTitle("LoadingHTMLString", for: UIControl.State.normal)
        buttonLoadHTMLString.frame = CGRect(x:0, y:0, width:117, height:30)
        // 指定事件处理方法
        buttonLoadHTMLString.addTarget(self, action: #selector(testLoadHTMLString), for: UIControl.Event.touchUpInside)
        buttonBar.addSubview(buttonLoadHTMLString)
        
        // 2. 添加LoadData按钮
        let buttonLoadData = UIButton(type: UIButton.ButtonType.system)
        buttonLoadData.setTitle("LoadData", for: UIControl.State.normal)
        buttonLoadData.frame = CGRect(x:137, y:0, width:67, height: 30)
        // 指定事件处理方法
        buttonLoadData.addTarget(self, action: #selector(testLoadData), for: UIControl.Event.touchUpInside)
        buttonBar.addSubview(buttonLoadData)
        
        // 3. 添加LoadRequest按钮
        let buttonLoadRequest = UIButton(type: UIButton.ButtonType.system)
        buttonLoadRequest.setTitle("LoadRequest", for: UIControl.State.normal)
        buttonLoadRequest.frame = CGRect(x:224, y:0, width: 92, height: 30)
        // 指定事件处理方法
        buttonLoadRequest.addTarget(self, action: #selector(testLoadRequest), for: UIControl.Event.touchUpInside)
        buttonBar.addSubview(buttonLoadRequest)
        
        // 4. 添加WKWebView
        self.webView = WKWebView(frame: CGRect(x:0, y:60, width:screen.size.width, height:(screen.size.height - 80)))
        self.view.addSubview(self.webView)
    }
    
    @IBAction func testLoadHTMLString(sender: AnyObject) {
        let htmlPath = Bundle.main.path(forResource: "index", ofType: "html")
        let bundleUrl = NSURL.fileURL(withPath: Bundle.main.bundlePath)
        
        do {
            let html = try NSString(contentsOfFile: htmlPath!, encoding: String.Encoding.utf8.rawValue)
            self.webView.loadHTMLString(html as String, baseURL: bundleUrl)
        } catch let err as NSError {
            NSLog(err.description)
        }
        NSLog("testLoadHTMLString")
    }

    @IBAction func testLoadData(sender: AnyObject) {
        let htmlPath = Bundle.main.path(forResource: "index", ofType: "html")
        let bundleUrl = NSURL.fileURL(withPath: Bundle.main.bundlePath)
        let htmlData = NSData(contentsOfFile: htmlPath!)
        
        self.webView.load(htmlData! as Data, mimeType: "text/html", characterEncodingName: "UTF-8", baseURL: bundleUrl)
        
        NSLog("testLoadData")
    }
    
    @IBAction func testLoadRequest(sender: AnyObject) {
        let url = NSURL(string: "https://www.baidu.com")
        let request = NSURLRequest(url: url! as URL)
        self.webView.load(request as URLRequest)
        NSLog("testLoadRequest")
    }
    
    // 实现WKNavigationDelegate委托协议
    // 开始加载时调用
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        NSLog("开始加载")
    }
    
    // 当内容开始返回时调用
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
        NSLog("内容开始返回")
    }
    
    // 加载完成之后调用
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        NSLog("内容加载完成")
    }
    
    // 加载失败d时调用
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        NSLog("内容加载失败 error: %@", error.localizedDescription)
    }
}

